### PR TITLE
[Backport 1.0] perf(hgraph): reduce visited list memory

### PR DESCRIPTION
**Pull request by jac0626**
## Summary

Backports #2449 to the `1.0` branch.

- replace the per-node visitation generation array with generation-stamped bitmap words
- preserve lazy reset semantics while reducing visited-list scratch memory
- include boundary, wraparound, zero-capacity, and randomized differential coverage

## Validation

- `make fmt` with clang-format 15
- `git diff --check`
- compiled `src/utils/visited_list.cpp` and `src/utils/visited_list_test.cpp` with the CMake-generated compilation commands
- `make test CASE='[ut][VisitedList]'` was blocked before project compilation by external dependency download timeouts; CI will run the targeted and repository checks with its dependency cache

Backport of #2449.

### CODE REVIEW

**Review by LHT129 — APPROVED** [review]
LGTM

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The diff is identical to the original PR.

**Summary of the change:**
- Replaces the per-node `uint16_t` tag array with generation-stamped bitmap words (`uint64_t` words + `uint16_t` tags per word), reducing visited-list scratch memory by ~80%.
- Preserves lazy reset semantics via per-word generation tags.
- Adds comprehensive test coverage: word boundaries, repeated sets, memory usage verification, tag overflow, zero-capacity edge case, and randomized differential testing against a `std::vector<bool>` reference.

**Review notes:**
- The `Allocate` nullptr check noted by the bot is a pre-existing pattern in this codebase — the `DefaultAllocator` throws `std::bad_alloc` on failure, so this is not a regression.
- The backport is faithful to #2449 with no divergence.

No blocking issues found.

File: src/utils/visited_list.h [hunk 1]
```diff
@@ -27,43 +27,57 @@ class Allocator;
...
     [[nodiscard]] bool
     Get(const InnerIdType& id) {
-        return this->list_[id] == this->tag_;
+        const auto word_id = static_cast<uint64_t>(id) / kBitsPerWord;
+        const auto mask = WordType{1} << (static_cast<uint64_t>(id) % kBitsPerWord);
+        return this->tags_[word_id] == this->tag_ and (this->words_[word_id] & mask) != 0;
     }
 
     void
     Prefetch(const InnerIdType& id) {
```
Comment (LHT129) [inline src/utils/visited_list.h:58]: [suggestion] `Prefetch` only prefetches `tags_` but not `words_`. In the `Get` path, when the tag matches, `words_[word_id]` must also be accessed. On a hot search path, this can cause an additional cache miss for the words array.

Consider also prefetching the corresponding word:
```cpp
PrefetchLines(this->words_ + word_id, 64);
```

This is a suggestion — the current approach may be intentionally trading off prefetch bandwidth for the common case where the tag check fails (element not visited). If that is the design intent, feel free to disregard.

**Review by LHT129 — COMMENTED** [review]
Left one inline comment (prefetch suggestion).

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The generation-stamped bitmap approach correctly reduces visited-list memory from `max_size * sizeof(uint16_t)` to `word_count * (sizeof(uint64_t) + sizeof(uint16_t))`, which is a significant reduction for large graphs.

**Summary of changes:**
- `visited_list.h`: Replaced per-element `uint16_t` array with interleaved `WordType` (uint64_t) bitmap words and `TagType` (uint16_t) generation tags, preserving lazy reset semantics.
- `visited_list.cpp`: Updated constructor to allocate a single contiguous buffer for words + tags, updated destructor with null guard, and updated `Reset()` to clear only tags on overflow.
- `visited_list_test.cpp`: Added comprehensive tests covering word boundaries, repeated sets, memory usage calculation, tag overflow (65535 resets), zero-size safety, and randomized differential testing against a `std::vector<bool>` reference.

**Existing comments to address:**
1. (gemini-code-assist) `Allocate` return value is not checked for nullptr — consider adding an assertion or null check.
2. (LHT129) `Prefetch` only prefetches `tags_` but not `words_` — this is a valid observation; the current design intentionally trades off prefetch bandwidth for the common case where the tag check fails, but worth considering.

No additional issues found. The code is well-structured and the test coverage is thorough.

File: src/utils/visited_list.h [hunk 2]
```diff
@@ -27,43 +27,57 @@ class Allocator;
...
+        const auto mask = WordType{1} << (static_cast<uint64_t>(id) % kBitsPerWord);
+        if (this->tags_[word_id] != this->tag_) {
+            this->tags_[word_id] = this->tag_;
+            this->words_[word_id] = mask;
+        } else {
+            this->words_[word_id] |= mask;
+        }
     }
 
     [[nodiscard]] bool
```
Comment (LHT129) [inline src/utils/visited_list.h:50]: [suggestion] `Get` is a query method that does not modify any member state — it only reads `tags_`, `words_`, and `tag_`. It should be marked `const` so it can be called on const references and to clearly signal its read-only semantics.

```cpp
[[nodiscard]] bool
Get(const InnerIdType& id) const {
    const auto word_id = static_cast<uint64_t>(id) / kBitsPerWord;
    const auto mask = WordType{1} << (static_cast<uint64_t>(id) % kBitsPerWord);
    return this->tags_[word_id] == this->tag_ and (this->words_[word_id] & mask) != 0;
}
```

File: src/utils/visited_list.h (hunk 1 above)
Comment (LHT129) [inline src/utils/visited_list.h:58]: [suggestion] `Prefetch` is also a read-only helper that does not modify any member state. It should be marked `const` for consistency and to allow calling on const references.

```cpp
void
Prefetch(const InnerIdType& id) const {
    const auto word_id = static_cast<uint64_t>(id) / kBitsPerWord;
    PrefetchLines(this->tags_ + word_id, 64);
}
```

**Review by LHT129 — COMMENTED** [review]
Code review for PR #2452 backport of visited list memory optimization.

File: src/utils/visited_list.h [hunk 3]
```diff
@@ -27,43 +27,57 @@ class Allocator;
...
+    static constexpr uint64_t kBitsPerWord = sizeof(WordType) * 8;
 
 public:
     explicit VisitedList(InnerIdType max_size, Allocator* allocator);
     ~VisitedList() override;
 
     void
     Set(const InnerIdType& id) {
-        this->list_[id] = this->tag_;
+        const auto word_id = static_cast<uint64_t>(id) / kBitsPerWord;
```
Comment (LHT129) [inline src/utils/visited_list.h:40]: [suggestion] `Set` and `Get` do not guard against `word_count_ == 0` (i.e., `max_size == 0`). In the constructor, when `word_count_ == 0`, both `words_` and `tags_` remain `nullptr`. If `Set` or `Get` is subsequently called, `word_id` evaluates to `0` and dereferences a null pointer.

The `Reset()` method already handles this case with an explicit `word_count_ > 0` guard before the `memset`. Consider adding a similar early return or assertion in `Set` and `Get`:

```cpp
void
Set(const InnerIdType& id) {
    if (word_count_ == 0) return;
    // ... existing logic
}

[[nodiscard]] bool
Get(const InnerIdType& id) {
    if (word_count_ == 0) return false;
    // ... existing logic
}
```

Alternatively, if the contract guarantees that `Set`/`Get` are never called on a zero-size list, an assertion would make the invariant explicit.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The diff is identical in intent and the test coverage (word boundaries, zero-size, tag overflow, randomized differential) is thorough.

All substantive review points have already been raised in the existing inline comments — no additional issues found.

LGTM for the backport once the existing inline suggestions are addressed.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The change replaces the per-node visitation generation array with generation-stamped bitmap words, which significantly reduces visited-list scratch memory while preserving lazy reset semantics.

The existing review comments from LHT129 and gemini-code-assist already cover the substantive issues:
- Null-pointer check after `Allocate` (gemini-code-assist)
- `Get` and `Prefetch` should be `const` (LHT129)
- `Prefetch` only prefetches `tags_` but not `words_` (LHT129)
- `Set`/`Get` missing `word_count_ == 0` guard for zero-size lists (LHT129)

The test coverage is thorough, including boundary conditions, tag overflow, zero-size lists, and randomized differential testing. No additional issues found.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The core change — replacing per-node `uint16_t` array with generation-stamped bitmap words — is correctly implemented.

There are 5 existing inline review comments that remain unaddressed, covering:
1. Null-pointer check after `Allocate()` (by gemini-code-assist)
2. `Prefetch` should also prefetch `words_` (by LHT129)
3. `Get` should be marked `const` (by LHT129)
4. `Prefetch` should be marked `const` (by LHT129)
5. `Set`/`Get` should guard against `word_count_ == 0` (by LHT129)

These are all valid suggestions. Please address them before merging.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The change replaces the per-node visitation array with a generation-stamped bitmap approach, which reduces memory usage while preserving lazy reset semantics.

**Summary of changes:**
- `visited_list.h`: Replaced `VisitedListType* list_` with `WordType* words_` + `TagType* tags_` bitmap representation. `Set`/`Get`/`Prefetch` updated to use word-level indexing with bitmask operations.
- `visited_list.cpp`: Constructor allocates a single buffer for both words and tags arrays. `Reset()` handles tag overflow with a full tags memset. Destructor guards against null pointer deallocation.
- `visited_list_test.cpp`: Added boundary tests, word-boundary tests, memory usage verification, tag overflow test, zero-size test with `CountingAllocator`, and a randomized differential test against a `std::vector<bool>` reference.

**Existing review comments** (from LHT129 and gemini-code-assist) have already identified the key areas worth attention:
1. Null-pointer check on `Allocate` return value
2. `Get` and `Prefetch` should be marked `const`
3. `Prefetch` could also prefetch the words array
4. `Set`/`Get` lack guard against `word_count_ == 0`

The code is well-structured, the tests are thorough, and the backport is faithful to the original. No additional blocking issues found.

**Review by LHT129 — COMMENTED** [review]
I have completed a thorough review of this backport PR. The core change — replacing the per-node `uint16_t` visitation array with generation-stamped bitmap words — is well-implemented and the logic is correct.

**Summary of findings:**

The PR has 5 existing unresolved review comments that cover the substantive issues:

1. **Null pointer check after `Allocate`** (gemini-code-assist): `allocator_->Allocate` return value is not checked before pointer arithmetic and `memset`. Consider adding a null check or assertion.

2. **`Get` should be `const`** (LHT129): `Get` is a read-only query method and should be marked `const`.

3. **`Prefetch` should be `const`** (LHT129): Same reasoning — read-only helper, should be `const`.

4. **`Prefetch` only prefetches `tags_` but not `words_`** (LHT129): On the hot search path, when the tag matches, `words_[word_id]` must also be accessed, causing a potential cache miss.

5. **`Set`/`Get` lack `word_count_ == 0` guard** (LHT129): When `word_count_ == 0`, `words_` and `tags_` are `nullptr`, but `Set`/`Get` will dereference them. The `Reset()` method already handles this case correctly.

**What looks good:**

- The bitmap word + generation tag design correctly preserves lazy reset semantics while significantly reducing memory (from 2 bytes per element to ~2.125 bytes per 64 elements).
- The constructor correctly handles `word_count_ == 0` (zero-capacity lists) by early-returning without allocation.
- The destructor correctly guards deallocation with a `nullptr` check.
- `Reset()` correctly handles tag overflow by clearing tags and resetting the generation counter.
- Test coverage is thorough: boundary conditions (word boundaries at 0, 63, 64, 65), tag overflow, zero-size lists, memory usage verification, and randomized differential testing against a `vector<bool>` reference.

No new issues found beyond what has already been raised. The existing comments should be addressed before merging.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The diff is identical to the original PR.

**Summary of the change:** Replaces the per-node visitation generation array with generation-stamped bitmap words, reducing visited-list scratch memory while preserving lazy reset semantics.

**Code quality:** The implementation is correct and well-structured. The test coverage is comprehensive, including boundary conditions, tag overflow, zero-size edge cases, and randomized differential testing against a reference implementation.

**Existing review comments worth addressing:**
- `Set`/`Get` do not guard against `word_count_ == 0` (null pointer dereference risk when `max_size == 0`)
- `Get` and `Prefetch` should be marked `const` since they are read-only methods
- Consider adding a null check after `Allocate` call in the constructor
- `Prefetch` only prefetches `tags_` but not `words_` — may cause additional cache miss in the hot path

No new issues found beyond what has already been flagged by other reviewers.

**Review by LHT129 — COMMENTED** [review]
I have completed a thorough review of this backport PR. The core change — replacing the per-node visitation array with generation-stamped bitmap words — is correctly implemented.

**Summary of findings:**

The existing inline comments from previous reviewers already cover the notable points:

1. **Null-pointer guard for `Allocate` return** (by @gemini-code-assist): The constructor does not check whether `allocator_->Allocate` returns `nullptr`. If the allocator can return null on OOM, this would lead to UB.

2. **`word_count_ == 0` guard in `Set`/`Get`** (by @LHT129): When `max_size == 0`, `words_` and `tags_` remain `nullptr`, but `Set`/`Get` would dereference them. The `Reset()` method already handles this case correctly.

3. **`const` qualifier on `Get` and `Prefetch`** (by @LHT129): Both methods are read-only and should be marked `const`.

4. **`Prefetch` only prefetches `tags_`** (by @LHT129): On a tag-match path, `words_[word_id]` must also be accessed, potentially causing an additional cache miss.

**On the positive side:**

- The bitmap word design with lazy tag reset is sound and correctly preserves the original lazy-reset semantics.
- The single-allocation layout (words + tags in one buffer) is efficient.
- The test coverage is comprehensive: basic set/get/reset, word boundaries, repeated sets, memory usage, tag overflow, zero-size edge case, and randomized differential testing against a `std::vector<bool>` reference.
- The `Reset()` method correctly handles the zero-size case with a `word_count_ > 0` guard.
- The destructor correctly guards against null `words_` before deallocation.

No new issues found beyond what has already been raised. The backport looks clean.

File: src/utils/visited_list.cpp [hunk 4]
```diff
@@ -15,27 +15,32 @@
```
Comment (LHT129) [inline src/utils/visited_list.cpp:15]: [critical] The constructor no longer guards against `max_size == 0` (the `word_count_ == 0` early return was removed). When `max_size == 0`:

1. `word_count_` is `0`, so `words_bytes` and `tags_bytes` are `0`.
2. `allocator_->Allocate(0)` is called — if it returns `nullptr`, then `words_` and `tags_` are `nullptr`.
3. `memset(tags_, 0, 0)` is called with a potentially null pointer — undefined behavior per the C standard.
4. The destructor calls `allocator_->Deallocate(words_)` without a null check — also problematic if `words_` is null.
5. `Reset()` calls `memset(tags_, 0, 0)` which suffers from the same UB.

The earlier version of this PR had a `word_count_ == 0` early return in the constructor and a `words_ != nullptr` guard in the destructor. Consider restoring those guards, or alternatively, ensure the allocator contract guarantees a non-null return for zero-size allocations and document that assumption.

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The generation-stamped bitmap approach is correctly implemented and the test coverage (word boundaries, memory usage, tag overflow) is solid.

I left one inline comment about a regression in the latest commit: the `word_count_ == 0` early return was removed from the constructor and the `words_ != nullptr` guard was removed from the destructor. When `max_size == 0`, this can lead to null pointer dereference in `memset` and `Deallocate`. The earlier version of this PR had these guards — consider restoring them.

The other existing inline comments from previous reviews (const-correctness for `Get`/`Prefetch`, prefetching `words_` in `Prefetch`, null guard in `Set`/`Get`) are still relevant and worth addressing.

**Review by LHT129 — COMMENTED** [review]
Thank you for this backport. The changes look clean and well-structured — the generation-stamped bitmap approach is a solid memory optimization.

I have reviewed the current diff and the existing inline comments already cover the notable points (null-pointer guards in `Set`/`Get`, `const`-correctness for `Get`/`Prefetch`, prefetch coverage for `words_`, and allocator null-return handling). I have no additional concerns to raise beyond what has already been discussed.

The test coverage is thorough — the randomized differential test, boundary tests, zero-size test, and tag overflow test all provide good confidence in the correctness of the implementation.

File: src/utils/visited_list.h (hunk 1 above)
Comment (LHT129) [inline src/utils/visited_list.h:58]: [suggestion] `Prefetch` also lacks a guard against `word_count_ == 0` (i.e., `max_size == 0`). When `word_count_ == 0`, `tags_` is `nullptr`, and `PrefetchLines(nullptr + 0, 64)` may trigger undefined behavior depending on the prefetch implementation.

This is the same class of issue already noted for `Set` and `Get`. Consider adding the same guard:

```cpp
void
Prefetch(const InnerIdType& id) {
    if (word_count_ == 0) return;
    const auto word_id = static_cast<uint64_t>(id) / kBitsPerWord;
    PrefetchLines(this->tags_ + word_id, 64);
}
```

**Review by LHT129 — COMMENTED** [review]
## Review Summary

This is a backport of #2449 to the `1.0` branch that replaces the per-node visitation generation array with generation-stamped bitmap words, reducing visited-list scratch memory.

### Current State

The current code (SHA `85183bec`) has addressed several issues raised in earlier review rounds:
- Constructor has `word_count_ == 0` early return to handle zero-size lists
- Destructor has `words_ != nullptr` guard before deallocation
- `Reset()` has `word_count_ > 0` guard before `memset`

### Outstanding Suggestions (from prior reviews)

Several suggestions from earlier reviews remain open for the author to consider:
1. `Set`/`Get`/`Prefetch` lack `word_count_ == 0` guards (LHT129 comments)
2. `Get` and `Prefetch` could be marked `const` (LHT129 comments)
3. `Prefetch` could also prefetch `words_` for better cache behavior (LHT129 comment)

### Test Coverage

The test additions are thorough, covering word boundaries, repeated sets, memory usage, tag overflow, zero-size lists, and randomized differential testing against a `std::vector<bool>` reference. Good coverage.

No new blocking issues found in this review round.

File: src/utils/visited_list_test.cpp [hunk 5]
```diff
@@ -58,6 +81,87 @@ TEST_CASE("VisitedList Basic Test", "[ut][VisitedList]") {
...
+                                                                  sizeof(VisitedList::TagType));
+        REQUIRE(vl_ptr->GetMemoryUsage() == expected);
+    }
+
+    SECTION("test tag overflow") {
+        vl_ptr->Set(0);
+        for (uint64_t i = 0; i < std::numeric_limits<VisitedList::TagType>::max(); ++i) {
+            vl_ptr->Reset();
+        }
+        REQUIRE_FALSE(vl_ptr->Get(0));
```
Comment (LHT129) [inline src/utils/visited_list_test.cpp:117]: [suggestion] test comment

Comment (LHT129) [inline src/utils/visited_list_test.cpp:117]: [suggestion] The `VisitedList Zero Size Test` validates `Reset()` and `GetMemoryUsage()` for a zero-size list, but does not exercise `Set()`, `Get()`, or `Prefetch()` when `max_size == 0`. Since the constructor allows `word_count_ == 0` (early return without allocating), these methods will dereference null `tags_`/`words_` pointers if called.

If zero-size lists are a valid use case, the test should cover these paths to document the expected behavior (e.g., no-op or assertion). If they are not valid, consider adding assertions in `Set`/`Get`/`Prefetch` to make the invariant explicit.

Suggested addition to the zero-size test:
```cpp
// Verify Set/Get/Prefetch are safe no-ops on zero-size list
// (or assert if they should never be called)
visited_list.Prefetch(0);
visited_list.Set(0);
REQUIRE_FALSE(visited_list.Get(0));
```

**Review by LHT129 — COMMENTED** [review]
This is a clean backport of #2449 to the `1.0` branch. The core change replaces a per-element `uint16_t` visited array with a generation-stamped bitmap using `uint64_t` words, which significantly reduces memory usage (from `2 * N` bytes to `(8 + 2) * ceil(N/64)` bytes).

The existing inline comments from previous reviews have already covered the notable points:

1. **Null-pointer guard for `Allocate`** — `allocator_->Allocate` may return `nullptr` on OOM.
2. **`word_count_ == 0` guards in `Set`/`Get`/`Prefetch`** — when `max_size == 0`, these methods dereference null pointers.
3. **`const` qualification for `Get` and `Prefetch`** — both are read-only and should be marked `const`.
4. **`Prefetch` only prefetches `tags_`** — consider also prefetching `words_` for the hot path where the tag matches.
5. **Zero-size test coverage** — the zero-size test does not exercise `Set`/`Get`/`Prefetch`.

The test coverage is solid with boundary, overflow, zero-size, and randomized differential tests. No new blocking issues found beyond what has already been raised.

**Review by LHT129 — COMMENTED** [review]
## Review Summary

This is a clean backport of #2449 to the `1.0` branch. The change replaces the per-node `uint16_t` visitation array with a generation-stamped bitmap approach, reducing visited-list memory from `max_size * sizeof(uint16_t)` to `word_count * (sizeof(uint64_t) + sizeof(uint16_t))`.

The implementation is correct and the test coverage is thorough (word boundaries, zero-size, tag overflow, randomized differential testing).

I have reviewed the current code against the existing inline comments. Most issues raised in prior reviews remain unaddressed — in particular:

- `Get` and `Prefetch` should be marked `const`
- `Set`, `Get`, and `Prefetch` lack guards against `word_count_ == 0`
- `Prefetch` only prefetches `tags_` but not `words_`
- `Allocate` return value is not checked for `nullptr`

No new issues found beyond what has already been reported.
